addon/PreviewProvider.js

```diff
@@ -359,15 +359,34 @@ PreviewProvider.prototype = {
   },
 
   /**
-   * Do some pre-processing on the links before inserting them into the metadata
+   * Do some pre-processing on the link before inserting it into the metadata
    * DB and adding them to the metadata cache
    */
-  processAndInsertMetadata(links) {
-    const processedLinks = this._processLinks(links);
-    this.insertMetadata(processedLinks);
+  processAndInsertMetadata(link, metadataSource) {
+    const processedLink = this._processLinks(link);
+    this.insertMetadata(processedLink, metadataSource);
   },
 
   /**
+   * Check if a single link exists in the metadata DB
+   */
+  asyncDoesSingleLinkExist: Task.async(function*(url) {
+    let link = this._processLinks([{url}])[0];
+    if (!link) {
+      return false;
+    }
+    // Hit the cache first
+    const key = link.cache_key;
+    const metadataFromCache = MetadataCache.cache.get(key);
+
+    if (!metadataFromCache) {
+      const linksMetadata = yield this._metadataStore.asyncExecuteQuery(`SELECT * FROM page_metadata WHERE cache_key = '${key}'`);
```

Comment:
I can see how it's not as efficient as it's doing a bunch of unnecessary work, but calling the existing method is more maintainable. This very well could be a performance critical section of code, and there are things we can do to optimize, but if it's not, then we'll have some quirky code exposure going on.

At the very least, the `SELECT` should be done from `MetadataStore.js` to encapsulate that logic instead of having it in `PreviewProvier` as well as the related tests in `test-PreviewProvider`. A simple fix could be to allow `asyncGetMetadataByCacheKey` to have an option of fetching images or not, so in our use here, it would only do `SELECT .. IN ..` Alternatively, a separate method for just `asyncCacheKeyExists` that takes a cache key is much easier to maintain and test than an arbitrary SQL querying API.